### DISCUSSION TITLE
Arduino Due + CNC board Protoneer for laser

### DISCUSSION THREAD

**Discussion by WZdunski** [terjeio/grblHAL#350] [General]
Hi, 
Do any of you have such a set? Arduino Due + CNC board Protoneer v3. grblHal >VER:1.1f.20211226
**Does anyone have $$ settings for this for laser?**
Exactly I am trying to set up the motors, nothing has changed in my kit, except for replacing the board with the Arduino Duo + CNC Protoneer v3. The settings from the previous board do not match.

When I move the X-axis 1mm in the program, the head moves about 15-16mm.
Also, when I have a slow movement, e.g. 100 mm/min or lower, the shift is not smooth, delicate, only the head shakes/jumps.
On a different board, everything runs smoothly, so it is not the fault of the construction.

- I using the XY axis ( Y dual by copy onboard CNC), polarization is correct. 

I tested $13, $n=G21 it doesn't work.

Please look on my settings, my someone has any idea? 

GrblHAL 1.1f ['$' or '$HELP' for help]
>G21:ok
[VER:1.1f.20211226:]
[OPT:VNMSLW,35,1024,3,0]
[NEWOPT:ENUMS,RT+,HOME,TC,SED]
[FIRMWARE:grblHAL]
[NVS STORAGE:*FLASH]
[DRIVER:SAM3X8E]
[DRIVER VERSION:211213]
[BOARD:Protoneer v3]
[G54:7.100,0.000,0.000]
[G55:0.000,0.000,0.000]
[G56:0.000,0.000,0.000]
[G57:0.000,0.000,0.000]
[G58:0.000,0.000,0.000]
[G59:0.000,0.000,0.000]
[G59.1:0.000,0.000,0.000]
[G59.2:0.000,0.000,0.000]
[G59.3:0.000,0.000,0.000]
[G28:0.000,0.000,0.000]
[G30:0.000,0.000,0.000]
[G92:0.000,0.000,0.000]
[HOME:0.000,0.000,0.000:0]
[TLO:0.000,0.000,0.000]
[PRB:0.000,0.000,0.000:0]
_______________________________
$$
$0=10.0
$1=25
$2=0
$3=0
$4=1
$5=0
$6=0
$7=0
$10=3
$11=0.020
$12=0.020
$13=0
$14=1
$15=0
$16=0
$17=0
$18=0
$19=0
$20=0
$21=0
$22=1
$23=0
$24=100.0
$25=1000.0
$26=250
$27=1.000
$28=0.100
$29=0.0
$30=500.000
$31=0.000
$32=1
$33=5000.0
$34=0.0
$35=0.0
$36=100.0
$37=0
$39=1
$40=0
$43=1
$44=4
$45=3
$46=0
$62=0
$63=3
$64=0
$65=0
$100=80.000
$101=80.000
$102=80.000
$110=10000.000
$111=10000.000
$112=10000.000
$120=800.000
$121=800.000
$122=800.000
$130=400.000
$131=300.000
$132=10.000
$341=0
$342=30.0
$343=25.0
$344=200.0
$345=100.0
$384=1

![Screenshot 2022-01-06 at 23 22 25](https://user-images.githubusercontent.com/96832902/148466307-189878d8-6d8f-44d6-a40b-e5fe93fa99ec.png)

**Comment by phil-barrett** [comment]
Are your steps/mm the same as the previous cnc board?  How are your stepper drivers set up for microstepping? If the difference is 16 to 1, then maybe you need to set the stepper drivers to 16:1 microstepping (or 3200 steps/rev).  The roughness of jogging makes me think that is the case.

**Reply by WZdunski** [reply]
@phil-barrett Hi, thx! Yes, you had right..

**Comment by WZdunski** [comment]
Question, the Arduino Due Atmel SAM3X8E ARM has two USBs, one NativeUSBPort manages the CPU SAM3X8E ARM 32bit and the other USB Programming Port manages CPU ATmega16U2 AVR 8bit.

Why is USB Native not working with GRBLHAL?

<img width="672" alt="Screenshot 2022-01-10 at 02 30 27" src="https://user-images.githubusercontent.com/96832902/148712037-b2768491-da67-479f-a3f8-1e9b0c3c2b29.png">

**Reply by terjeio** [reply]
Because you have not enabled it in in _my_machine.h_?

```
// Configuration
// Uncomment to enable.

#define USB_SERIAL_CDC     1 // Use native USB port for communication.
```

**Reply by WZdunski** [reply]
@terjeio HI, if I change that option is no communication on both USBs. If I off "#define USB_SERIAL_CDC     1 // Use native USB port for communication." I communicate by USB Programming port only.. 


[VER:1.1f.20211226:]
[OPT:VNMSL,35,1024,3,0]
Target buffer size found
[NEWOPT:ENUMS,RT+,TC,SED]
[FIRMWARE:grblHAL]
[NVS STORAGE:*FLASH]
[DRIVER:SAM3X8E]
[DRIVER VERSION:211213]
[BOARD:Protoneer v3]

**Reply by terjeio** [reply]
Seen from the processor there is only one USB port, the programming port is an UART port. Have you installed [a driver](https://www.arduino.cc/en/pmwiki.php?n=Guide/ArduinoDue) for the native USB port? This is what it looks like for me (on Win10):

![image](https://user-images.githubusercontent.com/20260062/149010635-35cd42b7-e5a0-4128-b871-37a82603feba.png)

**Reply by WZdunski** [reply]
@terjeio Yes, I did. Speed 115200
![Screenshot 2022-01-11 at 19 52 44](https://user-images.githubusercontent.com/96832902/149012895-883b85b2-7524-44f8-a8c1-e87cbfa28239.png)

**Reply by terjeio** [reply]
What do kind of response do you get when connecting to the port then? For me it works perfectly fine with both puTTY and [ioSender](https://github.com/terjeio/ioSender) - but starts up in Alarm state when no pins are connected (with default settings).

**Reply by WZdunski** [reply]
@terjeio So,LightBurn, I have "waiting for connection" on both ports USB in the board.

ioSender communicates with the board on NATIVE USB but on USB Programming port I have " No response received from  controller, exiting". 

**I using The LightBurn to work :) so, can you tell me what I have to change to have communication on LightBurn?**

**Reply by terjeio** [reply]
> I using The LightBurn to work :) so, can you tell me what I have to change to have communication on LightBurn?

To be honest I do not know since communication works for me but not for you for some reason.

You may try with adding
`#define USB_SERIAL_WAIT 1`
to _my_machine.h_ and see if that changes anything.

When configured to wait connecting with puTTY outputs these messages on the _first connect_ after a reset or power up:

![image](https://user-images.githubusercontent.com/20260062/149023860-697dbfbe-944d-49f6-a323-83ee4e639f10.png)

You may also try with the version uploaded today as I have been modifying the serial stream handling code recently - your issue _could_ be due to a regression.

**Reply by WZdunski** [reply]
@terjeio Unfortunately it doesn't work at LightBurn at all, I tried with "#define USB_SERIAL_WAIT 1" and without, I tried loading to Native and to Programmer USB, and all test was with the new version GRBLHAL. 

ok
[VER:1.1f.20220111:]
[OPT:VNMSL,35,1024,3,0]
Target buffer size found
[NEWOPT:ENUMS,RT+,TC,SED]
[FIRMWARE:grblHAL]
[NVS STORAGE:*FLASH]
[DRIVER:SAM3X8E]
[DRIVER VERSION:220111]
[BOARD:Protoneer v3]
ok

**Reply by terjeio** [reply]
I have [uploaded the binary](http://www.io-engineering.com/downloads) that works for me, try with that.

**Reply by WZdunski** [reply]
@terjeio I upload by BOSSA .bin file the grblHAL_Due_Native_USB.ino.zip and the LightBurn, ioSender not talking with the board on both USBs :(
Native USB is calls Bossa Programmer in Win.  

![Screenshot 2022-01-12 at 13 18 59](https://user-images.githubusercontent.com/96832902/149147850-a12a343a-4a72-4adc-84be-de5568bee620.png)

**Reply by WZdunski** [reply]
I think is some problem with LIGHBURN, because the ioSander, OpenBuildsCONTROL, Arduino IDE that talking to each other >> with parameter ON "#define USB_SERIAL_CDC     1 // Use native USB port for communication."

**Reply by terjeio** [reply]
Now I am completely confused, you can connect via native USB with ioSender/OpenBuildsControl but not Lightburn when you compile with `#define USB_SERIAL_CDC 1` uncommented?
If so it may have luck with changing the [compatibility level](https://github.com/grblHAL/core/wiki/Compatibility-level) option.

Be aware when connecting via native USB the controller is _not_ reset (and cannot be as it will kill the connection!) like it is is when connecting via the programming port. If Lightburn waits for the welcome message on connect then native USB cannot be used unless you can get them to change that.

When enabling USB_SERIAL_CDC the programming port is disabled for communication, both cannot be active at the same time.

**Reply by WZdunski** [reply]
@terjeio HI, so I made a record from my screen. If you are feeling confused looked for that :P 
That is connected by USB NATIVE. The board started working but connecting is very difficult.
I connect USB to the computer > Bossa programmer is on beginning, the softwares is not responding, by button on board I reset and I got an Arduino Due USB but LightBurn still not talking > if I connected by any other software and disconnect, the LightBurn started working > That's weird 👍 
Link to record >>
https://drive.google.com/file/d/1WCw30iqjeNlC_LqgvSJJ-xHsAaobWaN0/view?usp=sharing

**Reply by terjeio** [reply]
Did you uncomment `#define USB_SERIAL_CDC 1` and reflash between attempting to connect to Com13 and Com15? If not then that explains why Com13 did not work as both the programming port and the native USB port cannot be active and accepting commands at the same time.

Yes, it is a bit strange that Lightburn can talk to the controller after another sender has connected. The only reason I can think of is that you have `#define USB_SERIAL_WAIT 1` in _my_machine.h_ and that Lightburn connects in a way that does not [exit the wait loop](https://github.com/grblHAL/SAM3X8E/blob/c7066966b602bee6a95b5364c19e4d9dbaf618c9/grblHAL_Due/src/usb_serial.cpp#L280-L282). The other sender does, and then Lightburn can connect.

Just to be sure, when using the programming port then Ligthburn can connect without issues?

**Comment by phil-barrett** [comment]
Lightbulb works fine with compat level 0. At least it does for the teensy 4.1.  Be sure to set it to Grbl. I'm away from my machine until next week so can't check but check your baud rate

**Reply by terjeio** [reply]
> but check your baud rate

Baud rate is usually irrelevant when connecting via native USB. Due transfer speed in check mode is ~70Kbytes/sec and I can connect with whatever baud rate I choose (just checked with PuTTY).
